src/core/pinned_memory_manager.cc

```diff
@@ -193,22 +225,87 @@ PinnedMemoryManager::Create(const Options& options)
     return Status::Success;
   }
 
-  void* buffer = nullptr;
+  instance_.reset(new PinnedMemoryManager());
+  if (options.host_policy_map_.empty()) {
+    void* buffer = nullptr;
 #ifdef TRITON_ENABLE_GPU
-  auto err = cudaHostAlloc(
-      &buffer, options.pinned_memory_pool_byte_size_, cudaHostAllocPortable);
-  if (err != cudaSuccess) {
-    buffer = nullptr;
-    LOG_WARNING << "Unable to allocate pinned system memory, pinned memory "
-                   "pool will not be available: "
-                << std::string(cudaGetErrorString(err));
+    auto err = cudaHostAlloc(
+        &buffer, options.pinned_memory_pool_byte_size_, cudaHostAllocPortable);
+    if (err != cudaSuccess) {
+      buffer = nullptr;
+      LOG_WARNING << "Unable to allocate pinned system memory, pinned memory "
+                     "pool will not be available: "
+                  << std::string(cudaGetErrorString(err));
+    } else {
+      LOG_INFO << "Pinned memory pool is created at '"
+               << PointerToString(buffer) << "' with size "
+               << options.pinned_memory_pool_byte_size_;
+    }
+#endif  // TRITON_ENABLE_GPU
+    instance_->AddPinnedMemoryBuffer(
+        std::shared_ptr<PinnedMemory>(
+            new PinnedMemory(buffer, options.pinned_memory_pool_byte_size_)),
+        0);
   } else {
-    LOG_INFO << "Pinned memory pool is created at '" << PointerToString(buffer)
-             << "' with size " << options.pinned_memory_pool_byte_size_;
-  }
+    // Create only one buffer / manager should be created for one node,
+    // and all associated devices should request memory from the shared manager
+    std::map<int32_t, std::string> numa_map;
+    for (const auto host_policy : options.host_policy_map_) {
+      const auto numa_it = host_policy.second.find("numa-node");
+      if (numa_it != host_policy.second.end()) {
+        int32_t numa_id;
+        if (ParseIntOption("Parsing NUMA node", numa_it->second, &numa_id)
+                .IsOk()) {
+          numa_map.emplace(numa_id, host_policy.first);
+        }
+      }
+    }
+    for (const auto node_policy : numa_map) {
+      auto status =
+          SetNumaMemoryPolicy(options.host_policy_map_.at(node_policy.second));
+      if (!status.IsOk()) {
+        LOG_WARNING << "Unable to allocate pinned system memory for NUMA node "
+                    << node_policy.first << ": " << status.AsString();
+        continue;
+      }
+      unsigned long node_mask;
+      status = GetNumaMemoryPolicyNodeMask(&node_mask);
+      if (!status.IsOk()) {
+        LOG_WARNING << "Unable to get NUMA node set for current thread: "
+                    << status.AsString();
+        continue;
+      }
+      void* buffer = nullptr;
+#ifdef TRITON_ENABLE_GPU
+      auto err = cudaHostAlloc(
+          &buffer, options.pinned_memory_pool_byte_size_,
```

Comment:
We need to update the documentation in main.cc around the host policy and pinned memory pool flag to state that with numa the pool size is allocated on each numa node.